### PR TITLE
Add kubernetes-based backend

### PR DESCRIPTION
**Pull request by micahhausler**
Signed-off-by: Micah Hausler <mhausler@amazon.com>

## Description

The new backend uses the Kubernetes CRD types in Tinkerbell core. Boots could use a bit of a refactor/organization in how the backends work, but for now I've tried to fit in to the existing structure. 

## Why is this needed

This is the boots side of work for tinkerbell/proposals#46

Fixes: #

## How Has This Been Tested?

Tested with real bare metal workers using an image built from this branch. I will add E2E tests in a follow up PR, or can add to this branch

## How are existing users impacted? What migration steps/scripts do we need?

No impact to existing users

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

File: packet/k8s_client.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	"github.com/tinkerbell/tink/pkg/controllers"
+	"github.com/tinkerbell/tink/pkg/convert"
+	"github.com/tinkerbell/tink/protos/workflow"
+	"google.golang.org/grpc"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type K8sClient struct {
```
Comment (jacobweinstock) [inline packet/k8s_client.go:18]: what do you think about changing `K8sClient` to `K8SClient`? 

_"Words in names that are initialisms or acronyms (e.g. "URL" or "NATO") have a consistent case."_ -- [ref](https://github.com/golang/go/wiki/CodeReviewComments#initialisms)

**Review by jacobweinstock — COMMENTED** [review]
Looking good. Had a few questions and suggestions.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -35,7 +35,7 @@ jobs:
       - name: Generate all files
         run: nix-shell --run 'make gen'
       - name: goimports
-        run: go get golang.org/x/tools/cmd/goimports && goimports -d . | (! grep .)
+        run: go install golang.org/x/tools/cmd/goimports && goimports -d . | (! grep .)
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply .github/workflows/ci.yaml:38]: added `@latest`

File: packet/k8s_client.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	clientFunc func() crclient.Client
+	logger     log.Logger
+	namespace  string
+}
+
+func NewK8sClient(logger log.Logger, k8sAPI, kubeconfig string) (*K8sClient, error) {
+	ccfg := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeconfig},
+		&clientcmd.ConfigOverrides{ClusterInfo: clientcmdapi.Cluster{Server: k8sAPI}})
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply packet/k8s_client.go:27]: Its optional, but mirrors the Kubernetes client options

File: packet/k8s_client.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	config, err := ccfg.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	namespace, _, _ := ccfg.Namespace()
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply packet/k8s_client.go:32]: fixed

File: packet/k8s_client.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	manager, err := controllers.NewManager(config, controllers.GetServerOptions())
+	if err != nil {
+		return nil, errors.WithStack(err)
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply packet/k8s_client.go:35]: This just helps debug logged errors with a stacktrace

File: packet/client.go [hunk 6]
```diff
@@ -16,25 +16,26 @@ import (
...
 
 type hardwareGetter interface {
 }
 
 type Client interface {
 	GetWorkflowsFromTink(context.Context, HardwareID) (*tw.WorkflowContextList, error)
-	DiscoverHardwareFromDHCP(ctx context.Context, mac net.HardwareAddr, giaddr net.IP, circuitID string) (Discovery, error)
 	ReportDiscovery(ctx context.Context, mac net.HardwareAddr, giaddr net.IP, circuitID string) (Discovery, error)
 
+	DiscoverHardwareFromDHCP(ctx context.Context, mac net.HardwareAddr, giaddr net.IP, circuitID string) (Discovery, error)
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply packet/client.go:32]: Fixed in the rebase

File: packet/k8s_client.go (hunk 1 above)
Comment (micahhausler) [reply packet/k8s_client.go:18]: K8s isn't so much an acronym as a shorthand

**Comment by mmlb** [comment]
@micahhausler can we wait until #258 is merged to get this in? I'd be a ok with doing the actual rebase and handling the changes necessary. Having to rebase #258  is quite the pain.

**Comment by mmlb** [comment]
I'm here to make good on my promise @micahhausler want me to rebase and fixup?

**Review by jacobweinstock — APPROVED** [review]
thanks @micahhausler , just one small request

File: client/kubernetes/hardware_finder.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	manager, err := controllers.NewManager(config, controllers.GetServerOptions())
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	go manager.Start(context.Background())
```
Comment (mmlb) [inline client/kubernetes/hardware_finder.go:47]: hmm this is surprising to me, guess my whole notion of k8s controllers is wrong here (**very** likely). Why does boots need to start a controller/manager to get data? Is it that boots will go through this locally started manager or something?

File: client/kubernetes/hardware_finder.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
```
Comment (mmlb) [inline client/kubernetes/hardware_finder.go:128]: is there a reason this isn't shared for both finders?

File: .gitignore [hunk 9]
```diff
@@ -6,3 +6,4 @@
 /cmd/boots/boots
 /cmd/boots/boots-*-*
 coverage.txt
+hack/tools/
```
Comment (mmlb) [inline .gitignore:9]: why is this here?

File: client/kubernetes/k8s_models_test.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+					Metadata: &v1alpha1.HardwareMetadata{
+						Instance: nil,
+					},
+				},
+				Status: v1alpha1.HardwareStatus{},
+			},
+			want: nil,
+		},
+		{
+			name: "nil metadata instance",
```
Comment (mmlb) [inline client/kubernetes/k8s_models_test.go:48]: dupe name

File: client/tinkerbell/hardware_finder.go [hunk 11]
```diff
@@ -82,20 +82,12 @@ func (f *HardwareFinder) ByMAC(ctx context.Context, mac net.HardwareAddr, _ net.
...
-	return false, nil
-}
-
 // WorkflowFinder is a type for finding if a hardware ID has active workflows.
 type WorkflowFinder struct {
 	wClient tinkworkflow.WorkflowServiceClient
 }
 
-// NewFinder returns a *Finder that satisfies client.Finder.
+// NewFinder returns a *WorkflowFinder that satisfies client.WorkflowFinder.
```
Comment (mmlb) [inline client/tinkerbell/hardware_finder.go:90]:
```suggestion
// NewWorkflowFinder returns a *WorkflowFinder that satisfies client.WorkflowFinder.
```

File: cmd/boots/dhcp.go [hunk 12]
```diff
@@ -83,7 +83,7 @@ func (d dhcpHandler) serveDHCP(w dhcp4.ReplyWriter, req *dhcp4.Packet) {
 
 	circuitID, err := getCircuitID(req)
 	if err != nil {
-		mainlog.With("mac", mac, "err", err).Info("error parsing option82")
+		mainlog.With("mac", mac).Error(err, "error parsing option82")
```
Comment (mmlb) [inline cmd/boots/dhcp.go:86]: lets just get rid of this whole branch here. Don't want to treat this as an error and no real point in logging this imo.

File: cmd/boots/main.go [hunk 13]
```diff
@@ -220,15 +225,17 @@ func main() {
 	}
 }
 
-// getWorkflowFinder returns a no-op workflow finder if tinkerbell is not the backend
-func getWorkflowFinder() (client.WorkflowFinder, error) {
+// getWorkflowFinder returns a no-op workflow finder if tinkerbell or kubernetes is not the backend
```
Comment (mmlb) [inline cmd/boots/main.go:228]: hmm I think standalone should be allowed too right @jacobweinstock ?

Comment (jacobweinstock) [reply cmd/boots/main.go:228]: ah yes. seems like standalone should be in there.

File: client/kubernetes/hardware_finder.go (hunk 7 above)
Comment (jacobweinstock) [reply client/kubernetes/hardware_finder.go:47]: hmmm...removing this goroutine from this function might be something to consider. The name suggests that its is a constructor and having a constructor start a Go routine seems a bit unusual. Also, Boots' [design philosophy](https://github.com/tinkerbell/boots/blob/main/docs/DESIGNPHILOSOPHY.md) encourages 
> "When you spawn goroutines, make it clear when - or whether - they exit. [2](https://github.com/golang/go/wiki/CodeReviewComments#goroutine-lifetimes)" 

additionally: https://github.com/golang/go/wiki/CodeReviewComments#synchronous-functions

Comment (displague) [reply client/kubernetes/hardware_finder.go:47]: I think I see what you were going for here, keeping the same interface between tinkerbell and Kubernetes modes.

Passing the funcs back to main() would seem reasonable, and then have main() run things, or register the things that need running and let some runner handle the functions.

And then I looked at `main()` (with all its initialization and go funcs).  

> Boots could use a bit of a refactor/organization in how the backends work, but for now I've tried to fit in to the existing structure.

yep.

Comment (micahhausler) [reply client/kubernetes/hardware_finder.go:47]: The manager includes a client-side indexed cache and I was just re-using what is in Tink. Given that we don't need a controller/manager, we can probably just instantiate a [`cluster.Cluster()`](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.2/pkg/cluster#Cluster)

File: cmd/boots/dhcp.go (hunk 12 above)
Comment (micahhausler) [reply cmd/boots/dhcp.go:86]: I don't know enough about this to feel comfortable changing the logic, I really just tried to standardize how the error logging worked as I debugged issues

File: cmd/boots/main.go (hunk 13 above)
Comment (micahhausler) [reply cmd/boots/main.go:228]: How does that apply? Standalone doesn't have a Tinkerbell backend right?

File: cmd/boots/main.go [hunk 14]
```diff
@@ -351,6 +360,8 @@ func newCLI(cfg *config, fs *flag.FlagSet) *ffcli.Command {
 	fs.StringVar(&cfg.dhcpAddr, "dhcp-addr", conf.BOOTPBind, "IP and port to listen on for DHCP.")
 	fs.StringVar(&cfg.syslogAddr, "syslog-addr", conf.SyslogBind, "IP and port to listen on for syslog messages.")
 	fs.StringVar(&cfg.extraKernelArgs, "extra-kernel-args", "", "Extra set of kernel args (k=v k=v) that are appended to the kernel cmdline when booting via iPXE.")
+	fs.StringVar(&cfg.kubeconfig, "kubeconfig", "", "The Kubernetes config file location. Only applies if DATA_MODEL_VERSION=kubernetes.")
+	fs.StringVar(&cfg.kubeAPI, "kubernetes", "", "The Kubernetes API URL. Only applies if DATA_MODEL_VERSION=kubernetes.")
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply cmd/boots/main.go:364]: Sure

**Comment by micahhausler** [comment]
> oh, also would you be able to provide some really basic instructions for use? maybe just how (or that its needed) to generate the crds from the tink repo and apply them to a cluster? or a link maybe to a guide for deploying Tink server/controller?

I'm going to hold of on E2E instructions for now, and do comprehensive documentation for all the components once the Hegel changes get in

**Review by jacobweinstock — APPROVED** [review]

File: cmd/boots/main.go (hunk 13 above)
Comment (mmlb) [reply cmd/boots/main.go]: standalone only applies to the hardware fetcher, workflows can still happen with standalone.

File: cmd/boots/dhcp.go (hunk 12 above)
Comment (mmlb) [reply cmd/boots/dhcp.go:86]: sounds fine.

File: cmd/boots/main.go (hunk 13 above)
Comment (micahhausler) [reply cmd/boots/main.go]: With which Tinkerbell data model? Tink on K8s or Tink on postgres?

Comment (micahhausler) [reply cmd/boots/main.go]: For backwards compatibility, I'll use Tink on Postgres. We'll have to figure how to transition this later

Comment (mmlb) [reply cmd/boots/main.go]: I wonder if standalone should also imply workflows from json too :thinking:

Comment (mmlb) [reply cmd/boots/main.go]: @jacobweinstock I take it you use standalone w/ workflows, thoughts?

**Review by jacobweinstock — DISMISSED** [review]

File: cmd/boots/main.go [hunk 15]
```diff
@@ -220,15 +221,52 @@ func main() {
...
-// getWorkflowFinder returns a no-op workflow finder if tinkerbell is not the backend
-func getWorkflowFinder() (client.WorkflowFinder, error) {
-	dataModelVersion := os.Getenv("DATA_MODEL_VERSION")
-	switch dataModelVersion {
+func getFinders(l log.Logger, c *config) (client.WorkflowFinder, client.HardwareFinder, error) {
+	var hf client.HardwareFinder
+	var wf client.WorkflowFinder = &client.NoOpWorkflowFinder{}
+	var err error
+
+	switch os.Getenv("DATA_MODEL_VERSION") {
```
Comment (chrisdoherty4) [inline cmd/boots/main.go:229]: Have we a 'todo' somewhere, or on someones list, to figure out wiring of configuration in Boots generally?

@jacobweinstock @micahhausler

File: cmd/boots/main.go [hunk 16]
```diff
@@ -351,6 +370,8 @@ func newCLI(cfg *config, fs *flag.FlagSet) *ffcli.Command {
 	fs.StringVar(&cfg.dhcpAddr, "dhcp-addr", conf.BOOTPBind, "IP and port to listen on for DHCP.")
 	fs.StringVar(&cfg.syslogAddr, "syslog-addr", conf.SyslogBind, "IP and port to listen on for syslog messages.")
 	fs.StringVar(&cfg.extraKernelArgs, "extra-kernel-args", "", "Extra set of kernel args (k=v k=v) that are appended to the kernel cmdline when booting via iPXE.")
+	fs.StringVar(&cfg.kubeconfig, "kubeconfig", "", "The Kubernetes config file location. Only applies if DATA_MODEL_VERSION=kubernetes.")
+	fs.StringVar(&cfg.kubeAPI, "kubernetes", "", "The Kubernetes API URL, used for in-cluster client construction. Only applies if DATA_MODEL_VERSION=kubernetes.")
```
Comment (chrisdoherty4) [inline cmd/boots/main.go:374]: I think this is different to Tink and Hegel. I think Tink and Hegel use `--kubernetes-api`?

**Comment by chrisdoherty4** [comment]
lgtm

**Review by chrisdoherty4 — APPROVED** [review]

File: cmd/boots/main.go (hunk 15 above)
Comment (micahhausler) [reply cmd/boots/main.go:229]: Created #618

File: cmd/boots/main.go (hunk 16 above)
Comment (chrisdoherty4) [reply cmd/boots/main.go:374]: Gotcha. I'll do the same on Hegel.

File: client/kubernetes/cluster.go [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+)
+
+var (
+	runtimescheme = runtime.NewScheme()
+)
+
+func init() {
+	_ = clientgoscheme.AddToScheme(runtimescheme)
+	_ = v1alpha1.AddToScheme(runtimescheme)
+}
```
Comment (mmlb) [inline client/kubernetes/cluster.go:28]: we're trying to minimize `init()s` can this be done some other way? sync.Once.Do() in NewCluster maybe?

File: cmd/boots/main.go (hunk 13 above)
Comment (jacobweinstock) [reply cmd/boots/main.go]: i use standalone mainly for testing, this seems fine regardless.

File: cmd/boots/main.go (hunk 16 above)
Comment (micahhausler) [reply cmd/boots/main.go:374]: Updated https://github.com/tinkerbell/tink/pull/619 with reasons

File: client/kubernetes/cluster.go (hunk 17 above)
Comment (micahhausler) [reply client/kubernetes/cluster.go]: Done

**Review by jacobweinstock — APPROVED** [review]

**Review by mmlb — CHANGES_REQUESTED** [review]
:tada:

**Review by mmlb — APPROVED** [review]

**Pull request merged**